pkg/addons/addons.go

```diff
@@ -58,27 +62,53 @@ func Ensure(s *state.State) error {
 		Config:      s.Cluster,
 		Credentials: creds,
 	}
-	if err := getManifests(s, templateData); err != nil {
-		return errors.WithStack(err)
+
+	addonsPath, dirs, err := traverseAddonsDirectory(s)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse the addons directory")
 	}
 
-	if err := applyAddons(s); err != nil {
-		return errors.Wrap(err, "failed to apply addons")
+	for _, addonDir := range dirs {
+		if len(addonDir) == 0 {
+			s.Logger.Info("Applying addons from the root directory...")
+		} else {
+			s.Logger.Infof("Applying addon %q...", addonDir)
+		}
+
+		manifest, err := getManifestsFromDirectory(s, templateData, addonsPath, addonDir)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		if len(strings.TrimSpace(manifest)) == 0 {
+			if len(addonDir) != 0 {
+				s.Logger.Warnf("Addon directory %q is empty, skipping...", addonDir)
+			}
+			continue
+		}
+
+		if err := applyAddons(s, manifest, addonDir); err != nil {
+			return errors.Wrap(err, "failed to apply addons")
+		}
 	}
 
 	return nil
 }
 
-func applyAddons(s *state.State) error {
-	return errors.Wrap(s.RunTaskOnLeader(runKubectl), "failed to apply addons")
-}
-
-func runKubectl(s *state.State, _ *kubeoneapi.HostConfig, _ ssh.Connection) error {
-	if err := s.Configuration.UploadTo(s.Runner.Conn, s.WorkDir); err != nil {
-		return errors.Wrap(err, "failed to upload manifests")
-	}
-	_, _, err := s.Runner.Run(fmt.Sprintf(kubectlApplyScript, addonLabel), runner.TemplateVariables{
-		"FILE_NAME": fmt.Sprintf("%s/addons/", s.WorkDir),
+func applyAddons(s *state.State, manifest string, addonName string) error {
+	return s.RunTaskOnLeader(func(s *state.State, _ *kubeoneapi.HostConfig, conn ssh.Connection) error {
+		var (
+			cmd            = fmt.Sprintf(kubectlApplyScript, addonLabel, addonName)
+			r              = strings.NewReader(manifest)
```

Comment:
```suggestion
			stdin              = strings.NewReader(manifest)
```

To make it clean the intention and role

Reply:
@kron4eg Fixed, PTAL.